docs_nnx/guides/flax_gspmd.md

```diff
@@ -0,0 +1,378 @@
+
+Note that with the correct sharding for all inputs, the output will be sharded in the most natural way even without `jax.jit`. See the example below - even without `jax.lax.with_sharding_constraint` on the output `y`, it was still sharded as `('data', None)`.
+
+> If you are interested in why: The second matmul of `DotReluDot.__call__` has two inputs of sharding `('data', 'model')` and `('model', None)`, in which both inputs' contraction axis are `model`. So a reduce-scatter matmul happened and will natually shard the output as `('data', None)`. Check out the [JAX shard map collective guide](https://jax.readthedocs.io/en/latest/notebooks/shard_map.html#example-2-psum-scatter-the-result) and its examples if you want to learn mathematically how it happened at low level.
+
+```{code-cell} ipython3
+# In data parallelism, the first dimension (batch) will be sharded on `data` axis
+data_sharding = NamedSharding(mesh, PartitionSpec('data', None))
+input = jax.device_put(jnp.ones((8, 1024)), data_sharding)
+
+with mesh:
+  output = sharded_model(input)
+print(output.shape)
+jax.debug.visualize_array_sharding(output)  # Also sharded as ('data', None)
+```
+
+Now the rest of the training loop is pretty conventional - almost the same as the example in [NNX Basics](https://flax-nnx.readthedocs.io/en/latest/nnx_basics.html#transforms), except that the inputs and labels are also explicitly sharded.
+
+`jax.jit` will adjust and automatically choose the best layout based on how its inputs are already sharded, so try out different shardings for your own model and inputs.
+
+```{code-cell} ipython3
+optimizer = nnx.Optimizer(sharded_model, optax.adam(1e-3))  # reference sharing
+
+@nnx.jit
+def train_step(model, optimizer, x, y):
+  def loss_fn(model: DotReluDot):
+    y_pred = model(x)
+    return jnp.mean((y_pred - y) ** 2)
+
+  loss, grads = nnx.value_and_grad(loss_fn)(model)
+  optimizer.update(grads)
+
+  return loss
+
+input = jax.device_put(jax.random.normal(jax.random.key(1), (8, 1024)), data_sharding)
+label = jax.device_put(jax.random.normal(jax.random.key(2), (8, 1024)), data_sharding)
+
+with mesh:
+  for i in range(5):
+    loss = train_step(sharded_model, optimizer, input, label)
+    print(loss)    # model (over-)fitting to the labels quickly
+```
+
+## Profiling
+
+If you are running on a TPU pod or a pod slice, you can use a custom `block_all` utility function, as defined below, to measure the performance:
+
+```{code-cell} ipython3
+%%timeit
+
+def block_all(xs):
+  jax.tree_util.tree_map(lambda x: x.block_until_ready(), xs)
+  return xs
+
+with mesh:
+  new_state = block_all(train_step(sharded_model, optimizer, input, label))
+```
+
+## Logical axis annotation
```

Comment:
Nice! I didn't know about `sharding_rules`. In `nnx_lm1b` with have this other pattern which maps the mesh axes in the constructor:
- https://github.com/google/flax/blob/e3772b2d01ba8aca7a09644ed66d4bc072c8782b/examples/lm1b_nnx/configs/default.py#L22
- https://github.com/google/flax/blob/e3772b2d01ba8aca7a09644ed66d4bc072c8782b/examples/lm1b_nnx/models.py#L214

Reply:
Yeah I added them recently to align with Linen's `LogicallyPartitioned`. It's just annotations so there's actually a ton of ways to make them work, and I like how you made it in `nnx_lm1b`!